### PR TITLE
[14.0] [mig] pos_user_restriction

### PR DESCRIPTION
**Pull request by hkapatel-initos**
Migration to 14.0.
In v12.0 there is no restricted group at the opening of session time and in version 14.0 system strictly allows to set a minimal access group of point of sale.
Only if there is a default user in version 14, the user will be set. The point of sale session is open
so select Allowed Employees.
Allowed Employees
Then assign "Authorized Employees"  Select the same user(cashier) that is in "Assigned POS Only" Use employee credentials to log in to the PoS session and switch cashier.

### CODE REVIEW

**Comment by hkapatel-initos** [comment]
@etobella It would be nice if you could review the PR.

File: pos_user_restriction/tests/test_pos_user_restriction.py [hunk 1]
```diff
@@ -41,31 +41,38 @@ def setUpClass(cls):
...
         # assigned_user_ids is set with pos_user: only pos_user can read
-        pos_configs = self.pos_config_model.sudo(self.pos_user.id).search([])
+        pos_configs = self.pos_config_model.with_user(self.pos_user.id).search([])
         self.assertTrue(pos_configs)
-        pos_configs = self.pos_config_model.sudo(self.pos_user_assigned_pos.id).search(
-            []
+        pos_configs = self.pos_config_model.with_user(
+            self.pos_user_assigned_pos.id
+        ).search([])
+        # self.assertFalse(pos_configs)
```
Comment (dsolanki-initos) [inline pos_user_restriction/tests/test_pos_user_restriction.py]: Why you have commented this line?

File: pos_user_restriction/security/ir.model.access.csv [hunk 2]
```diff
@@ -7,7 +7,8 @@ access_stock_warehouse_pos_user,stock.warehouse pos_user,stock.model_stock_wareh
 access_stock_move_pos_user,stock.move pos_user,stock.model_stock_move,group_assigned_points_of_sale_user,1,1,1,1
 access_report_pos_order,report.pos.order,point_of_sale.model_report_pos_order,group_assigned_points_of_sale_user,1,1,1,1
 access_account_journal_pos_user,account.journal pos_user,account.model_account_journal,group_assigned_points_of_sale_user,1,0,0,0
-access_account_bank_statement,account.bank.statement,account.model_account_bank_statement,group_assigned_points_of_sale_user,1,1,1,0
+access_account_payment_method_pos_user,account.payment.method pos_user,account.model_account_payment_method,group_assigned_points_of_sale_user,1,0,0,0
+access_account_bank_statement,account.bank.statement,account.model_account_bank_statement,group_assigned_points_of_sale_user,1,1,1,0Assigned users
```
Comment (dsolanki-initos) [inline pos_user_restriction/security/ir.model.access.csv]: can you please check  I think Assigned users added at the end of this line?

File: pos_user_restriction/security/pos_security.xml [hunk 3]
```diff
@@ -14,29 +15,29 @@
         />
     </record>
 
-<data noupdate="1">
+<data noupdate="0">
```
Comment (dsolanki-initos) [inline pos_user_restriction/security/pos_security.xml]: Why this is changed?

File: pos_user_restriction/models/pos_config.py [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
     _inherit = "pos.config"
+
     assigned_user_ids = fields.Many2many(
         "res.users",
         string="Assigned users",
         help="Restrict some users to only access their assigned points of sale. "
         "In order to apply the restriction, the user needs the "
         "'User: Assigned POS Only' group",
     )
+    set_pos_user_group = fields.Boolean(
```
Comment (dsolanki-initos) [inline pos_user_restriction/models/pos_config.py]: Why we need new field?

File: pos_user_restriction/tests/test_pos_user_restriction.py [hunk 5]
```diff
@@ -41,31 +41,38 @@ def setUpClass(cls):
...
-        self.assertTrue(pos_configs)
-        pos_configs = self.pos_config_model.sudo(self.pos_user_assigned_pos.id).search(
-            []
-        )
+        pos_configs = self.pos_config_model.with_user(self.pos_user.id).search([])
         self.assertTrue(pos_configs)
+        pos_configs = self.pos_config_model.with_user(
+            self.pos_user_assigned_pos.id
+        ).search([])
+        # self.assertTrue(pos_configs)
```
Comment (dsolanki-initos) [inline pos_user_restriction/tests/test_pos_user_restriction.py]: Same here please remove the comment from this line.

**Review by dsolanki-initos — COMMENTED** [review]
Functional review LGTM. Some comments that need to be resolved.

File: pos_user_restriction/models/pos_config.py (hunk 4 above)
Comment (hkapatel-initos) [reply pos_user_restriction/models/pos_config.py]: switch the cashier according to the code of the Owl Structure in  v14  for the switch cashier (employee) calls the (pos_user_group) method to switch cashier easily  and  can start a session so I have put a new field and method

File: pos_user_restriction/security/pos_security.xml (hunk 3 above)
Comment (hkapatel-initos) [reply pos_user_restriction/security/pos_security.xml]: sorry  I forgot it

**Comment by hkapatel-initos** [comment]
@dsolanki-initos that suggestions are implemented you can re-review it

**Comment by hkapatel-initos** [comment]
@chienandalu can you please add this to the 14.0 milestone and merge this  PR?

**Comment by hkapatel-initos** [comment]
@eLBati It would be nice if you could review the PR.

**Comment by eLBati** [comment]
@hkapatel-initos sorry I'm not using v14 at the moment so I am not able to evaluate the correctness of the changes

**Review by hkapatel-initos — COMMENTED** [review]
> @hkapatel-initos sorry I'm not using v14 at the moment so I am not able to evaluate the correctness of the changes

okay

**Comment by hkapatel-initos** [comment]
@legalsylvain It would be nice if you could review the PR.

**Comment by hkapatel-initos** [comment]
@ivantodorovich can you please review this PR?

File: pos_user_restriction/models/pos_config.py (hunk 4 above)
Comment (ivantodorovich) [reply pos_user_restriction/models/pos_config.py]: This is not ok.

Try making the `group_pos_user_id` field a computed field instead, it should probably work

Comment (hkapatel-initos) [reply pos_user_restriction/models/pos_config.py]: @ivantodorovich Thanks for the review suggestions are implemented you can re-review it.

File: pos_user_restriction/models/pos_config.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        "res.groups",
+        compute="_compute_set_pos_user_group",
+        string="Point of Sale User Group",
+        help="This field is there to pass"
+        "the id of the pos user group to the point of sale client.",
+        store=True,
+    )
+
+    @api.depends("assigned_user_ids")
+    def _compute_set_pos_user_group(self):
```
Comment (ivantodorovich) [inline pos_user_restriction/models/pos_config.py]:
```suggestion
    def _compute_group_pos_user_id(self):
```

As per OCA guidelines, use `_compute_<field_name>` for compute method names

File: pos_user_restriction/tests/test_pos_user_restriction.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        self.pos_config_main.assigned_user_ids = [(6, 0, [self.pos_user.id])]
+        # assigned_user_ids is set with pos_user: only pos_user can read
+        pos_configs = self.pos_config_model.with_user(self.pos_user.id).search([])
+        self.assertTrue(pos_configs)
+        pos_configs = self.pos_config_model.with_user(
+            self.pos_user_assigned_pos.id
+        ).search([])
+        self.assertFalse(pos_configs)
+
+        self.pos_config_main._compute_set_pos_user_group()
```
Comment (ivantodorovich) [inline pos_user_restriction/tests/test_pos_user_restriction.py]:
```suggestion
```
You can probably remove this line

**Review by ivantodorovich — APPROVED** [review]
Small comments otherwise LGTM

I'm pre-approving

**Comment by hkapatel-initos** [comment]
@HaraldPanten can u please review this pr ?

**Comment by hkapatel-initos** [comment]
@christophlsa can you please review this pr?

**Review by HaraldPanten — APPROVED** [review]
Functional review 👍

File: pos_user_restriction/models/pos_config.py (hunk 6 above)
Comment (ivantodorovich) [reply pos_user_restriction/models/pos_config.py]: Please check this comment as I think it hasn't been taken care of ☺️

**Comment by legalsylvain** [comment]
@hkapatel-initos can you answer to the @ivantodorovich remarks ? 
thanks.

File: pos_user_restriction/models/pos_config.py (hunk 6 above)
Comment (hkapatel-initos) [reply pos_user_restriction/models/pos_config.py]: @ivantodorovich that suggestions are implemented  can you  Re-check it

**Comment by hkapatel-initos** [comment]
> @hkapatel-initos can you answer to the @ivantodorovich remarks ?
> thanks.

Done

**Comment by ivantodorovich** [comment]
Thanks !! :tada: 

/ocabot merge nobump

**Pull request merged**